File: backend/core/services/communication_words.py
"""How many words a communication has, in either language (R48 item 7).

A communication assignment carries a `word_limit`, and the submit route
refused a memo over it by counting `len(content.split())`. Chinese is written
without spaces, so a Chinese memo of any length counted as one word and the
limit never bound for a Chinese team. The integrator's decision under R48:
count CJK letters at 1.5 letters per word, added to the space-split count of
the rest, so a Chinese team is limited like an English one.

This is the only place the server counts, and
`frontend/globalstrat-frontend/src/communicationWords.js` is the only place
the screen counts; the two carry the same letter ranges and the same integer
arithmetic, and their tests share one table of cases, so the live count a
student watches is the count the server refuses on.

The rule:

* a CJK letter is a Han ideograph (the unified block, extensions A and the
  supplementary-plane extensions, the compatibility ideographs, and the
  iteration marks 々〆〇), a Hiragana or Katakana letter (including the
  phonetic extensions and the half-width forms), or a Hangul syllable or
  jamo;
* every 1.5 CJK letters are one word, rounded up, so a single letter is one
  word and never zero: `ceil(letters * 2 / 3)`, in integers;
* CJK punctuation and full-width symbols (。，、「」！？ and the like) separate
  words and are not counted; full-width digits and Latin letters are words
  like their ASCII forms;
* everything else is split on whitespace, exactly as before.
"""
import re

CJK_LETTER = re.compile(
    '['
    '㐀-䶿'          # CJK Unified Ideographs Extension A
    '一-鿿'          # CJK Unified Ideographs
    '豈-﫿'          # CJK Compatibility Ideographs
    '\U00020000-\U0002fa1f'  # Extensions B..F and the supplementary compat block
    '々-〇'          # 々 〆 〇
    '぀-ゟ'          # Hiragana
    '゠-ヿ'          # Katakana
    'ㇰ-ㇿ'          # Katakana Phonetic Extensions
    'ｦ-ﾟ'          # Half-width Katakana
    'ᄀ-ᇿ'          # Hangul Jamo
    '㄰-㆏'          # Hangul Compatibility Jamo
    '가-힯'          # Hangul Syllables
    ']')

CJK_SEPARATOR = re.compile(
    '['
    '　-〄〈-〿'  # CJK symbols and punctuation, less 々〆〇
    '！-／：-＠'  # full-width punctuation, not digits/letters
    '［-｀｛-･'
    ']')

CJK_LETTERS_PER_WORD = 1.5


def count_words(text):
    """The number of words in `text` under the rule above; `None` is empty."""
    source = text or ''
    letters = len(CJK_LETTER.findall(source))
    rest = CJK_SEPARATOR.sub(' ', CJK_LETTER.sub(' ', source))
    # ceil(letters / 1.5) without a float, so the screen cannot disagree by a
    # rounding: 1 -> 1, 2 -> 2, 3 -> 2, 600 -> 400.
    cjk_words = (letters * 2 + 2) // 3
    return len(rest.split()) + cjk_words

File: frontend/globalstrat-frontend/src/communicationWords.js
/**
 * How many words a communication has, in either language (R48 item 7).
 *
 * The communications page shows a live "N / limit" count while a student
 * writes, and disables Submit past the limit. It counted by splitting on
 * spaces, so a Chinese memo counted as one word and the limit never bound
 * for a Chinese team. This is the screen's half of one rule; the server's
 * half is `backend/core/services/communication_words.py`, with the same
 * letter ranges and the same integer arithmetic, and the two tests share one
 * table of cases. Change both or neither.
 *
 * The rule:
 *  - a CJK letter is a Han ideograph (unified block, extension A, the
 *    supplementary-plane extensions, compatibility ideographs, and 々〆〇),
 *    a Hiragana or Katakana letter (with the phonetic extensions and the
 *    half-width forms), or a Hangul syllable or jamo;
 *  - every 1.5 CJK letters are one word, rounded up: ceil(letters * 2 / 3);
 *  - CJK punctuation and full-width symbols separate words and are not
 *    counted; full-width digits and Latin letters are words;
 *  - everything else is split on whitespace, as before.
 */

const CJK_LETTER = new RegExp(
  '['
  + '\\u3400-\\u4dbf'      // CJK Unified Ideographs Extension A
  + '\\u4e00-\\u9fff'      // CJK Unified Ideographs
  + '\\uf900-\\ufaff'      // CJK Compatibility Ideographs
  + '\\u{20000}-\\u{2fa1f}' // Extensions B..F and the supplementary compat block
  + '\\u3005-\\u3007'      // 々 〆 〇
  + '\\u3040-\\u309f'      // Hiragana
  + '\\u30a0-\\u30ff'      // Katakana
  + '\\u31f0-\\u31ff'      // Katakana Phonetic Extensions
  + '\\uff66-\\uff9f'      // Half-width Katakana
  + '\\u1100-\\u11ff'      // Hangul Jamo
  + '\\u3130-\\u318f'      // Hangul Compatibility Jamo
  + '\\uac00-\\ud7af'      // Hangul Syllables
  + ']',
  'gu',
);

const CJK_SEPARATOR = new RegExp(
  '['
  + '\\u3000-\\u3004\\u3008-\\u303f' // CJK symbols and punctuation, less 々〆〇
  + '\\uff01-\\uff0f\\uff1a-\\uff20' // full-width punctuation, not digits/letters
  + '\\uff3b-\\uff40\\uff5b-\\uff65'
  + ']',
  'gu',
);

export const CJK_LETTERS_PER_WORD = 1.5;

/** The number of words in `text` under the rule above; a missing value is 0. */
export function countWords(text) {
  const source = text || '';
  const letters = (source.match(CJK_LETTER) || []).length;
  const rest = source.replace(CJK_LETTER, ' ').replace(CJK_SEPARATOR, ' ').trim();
  const spaced = rest ? rest.split(/\s+/).length : 0;
  // ceil(letters / 1.5) in integers, as the server does it.
  const cjkWords = Math.floor((letters * 2 + 2) / 3);
  return spaced + cjkWords;
}
